### ISSUE TITLE
[Bug]: I updated my llama-index from 0.8.69.post2 to 0.9.10 and noticed memory growing over hours

### ISSUE THREAD

**Issue by alrooney** [run-llama/llama_index#9399]
### Bug Description

Memory usage is stable with 0.8.69.post2 it grows significantly ~300Mb -> 1Gb over hours with 0.9.10.  Here is a graph:
![image](https://github.com/run-llama/llama_index/assets/1009422/269b656c-8226-49e5-9841-19eafaa480d3)
The first deploy(c4248ec2) is going to 0.9.10 and the second deploy(bff4af04) is going back to 0.8.69.post2.  The steps likely represent a job I am running where I am periodically creating an index and sometimes adding documents but mostly not i.e. I am running this code every hour and sometimes (probably most of the time) documents is empty i.e. []
```
def _setup_index(collection_name: str,
                 documents: list = [],
                 temperature: float = 0.0,
                 model: str = "gpt-3.5-turbo"):
    storage_context, service_context, token_counter = \
        _setup_contexts(collection_name, temperature = temperature, model = model)

    index = VectorStoreIndex.from_documents(documents,
                                            storage_context = storage_context,
                                            service_context = service_context)
    return index, token_counter

def _setup_contexts(collection_name: str,
                    temperature: float,
                    model: str):
    # construct vector store
    vector_store = QdrantVectorStore(
        client=QDRANT_CLIENT,
        collection_name=collection_name,
    )

    token_counter = TokenCountingHandler()
    callback_manager = CallbackManager([token_counter])

    service_context = ServiceContext.from_defaults(callback_manager=callback_manager,
                                                   llm=OpenAI(model=model,
                                                              temperature=temperature))
```
Hope this helps track down the issue.

### Version

0.9.10

### Steps to Reproduce

Create VectorStoreIndex and monitor memory usage.  See code above for how I am creating the index.

**Comment by alrooney** [comment]
In case it helps here's the version of lanchain I am using too:
<img width="1336" alt="image" src="https://github.com/run-llama/llama_index/assets/1009422/152d6ee0-e04d-45a8-9f07-323ccd710338">
That diff represents the only changes in release in the ```bff4af04``` release in the graph.

**Comment by logan-markewich** [comment]
@alrooney hmmm weird. And this same code on 0.8.69.post1 didn't have this issue?

Is QDrant in memory, or is it hosted on another server?

**Comment by alrooney** [comment]
Hi @logan-markewich, To clarify 0.8.69.post2 has no memory issue but 0.9.10 does.  I have not tried newer releases since 0.9.10.  So likely something introduced in 0.9.x but have not done a binary search of other releases.  I am using cloud hosted QDRANT so not on the same server.  The only thing running on this server is llama-index.

**Comment by logan-markewich** [comment]
Ok, finally took some time to dig into this (this will be a long comment, it's a long issue lol)

## Test Script

Here's the script I used for testing. I swapped in weaviate since I don't have access to qdrant cloud.

```python
from llama_index import Document, VectorStoreIndex, ServiceContext, StorageContext
from llama_index.llms import OpenAI
from llama_index.callbacks import CallbackManager, TokenCountingHandler
from llama_index.vector_stores import WeaviateVectorStore

from memory_profiler import profile


@profile
def _setup_index(
    collection_name: str,
    documents: list = [],
    temperature: float = 0.0,
    model: str = "gpt-3.5-turbo",
):
    storage_context, service_context, token_counter = _setup_contexts(
        collection_name, temperature=temperature, model=model
    )

    index = VectorStoreIndex.from_documents(
        documents, storage_context=storage_context, service_context=service_context
    )
    return index, token_counter


def _setup_contexts(collection_name: str, temperature: float, model: str):
    import weaviate

    auth_config = weaviate.AuthApiKey(api_key="...")

    client = weaviate.Client(
        url="...",
        auth_client_secret=auth_config,
    )

    # construct vector store
    vector_store = WeaviateVectorStore(client, class_prefix=collection_name)

    token_counter = TokenCountingHandler()
    callback_manager = CallbackManager([token_counter])

    service_context = ServiceContext.from_defaults(
        callback_manager=callback_manager,
        llm=OpenAI(model=model, temperature=temperature),
    )
    storage_context = StorageContext.from_defaults(vector_store=vector_store)

    return storage_context, service_context, token_counter


@profile
def main():
    # measure the memory usage after each step
    import random
    import time

    documents = [Document.example()] * 20

    for i in range(0, 10):
        index, token_counter = _setup_index(
            "LlamaIndex",
            documents=documents,
            temperature=0.0,
            model="gpt-3.5-turbo",
        )
        time.sleep(30)


if __name__ == "__main__":
    main()

```

## Results

![image](https://github.com/run-llama/llama_index/assets/22285038/19153c76-36c8-49b7-a0e9-1a8b425d7d45)


## Conclusion

These results aren't really satisfying. But both runs seem to indicate similar patterns of memory usage, so not entirely sure how to debug further. Going to try removing/swapping a few pieces to see if the pattern remains (i.e. callback handler, vector store)

**Comment by logan-markewich** [comment]
LOL I just realized the last iteration (11) is the output of the profile on the `main()` function -- that's why it looks like it levels off

**Comment by alrooney** [comment]
@logan-markewich thanks for digging into that.  So seems both 0.8.69.post1 and 0.9.10 keep increasing in your testing and based on your last comment?  Would it be helpful for me to run that on my setup with qdrant cloud?

**Comment by logan-markewich** [comment]
@alrooney I think at this point I just need to try and use some profiling tools (tracemalloc?) To dig into the issue.

If you have an experience or favorite tools for figuring this kind of stuff out, I'm definitely all ears 😅

**Comment by alrooney** [comment]
@logan-markewich I haven't been able to reproduce the issue with mprof and the code above either so not sure if there is something missing vs the code that was running on heroku.  I'm going to try to narrow it down on my end too.  I don't really have a recommendation re memory profilers in python, but I'll see if I can narrow down the release where it was introduced too.

**Comment by logan-markewich** [comment]
That would be extremely helpful! Thank you @alrooney !

**Comment by alrooney** [comment]
@logan-markewich I've been testing with different versions and was unable to reproduce it with 0.9.10 again.  Even though I only changed llama-index and langchain in the pipfile it seems that other libs were updated in the lock file so likely one of those and trying to track that down.  But I wanted to get back to you as soon as I figured this out.  It does not seem like this is a llama-index.  So sorry for the wild goose chase.  I'll let you get back to the fantastic work you are doing on llama-index!
